### PR TITLE
OpenMP target offloading for random123

### PR DESCRIPTION
**Pull request by ohm314**
**Description**

A clean branch based on the experiments done during GPU hackathon implementing OpenMP offloading for
coreneuron random123, keeping the CUDA/OpenACC versions intact so they can coexist.

- [x] Currently building with extra flags to make sure Random123 doesn't break, should be checked and put into build system
- [x] Check if OpenMP offloading is properly working
- [x] Check if different build variants are functional (that's what the CI is for)
- [x] Profile implementation

Fixes # (issue)

**How to test this?**

```bash
cmake .. -G Ninja -DCORENRN_ENABLE_GPU=ON -DCORENRN_ENABLE_NMODL=ON \
    -DCORENRN_NMODL_FLAGS=""  -DCMAKE_CUDA_COMPILER=nvcc  \
    -DCMAKE_CXX_FLAGS="-DR123_USE_SSE=0" -DCMAKE_CXX_COMPILER=nvc++ \
    -DCORENRN_ENABLE_OPENMP=ON -DCORENRN_ENABLE_OPENMP_OFFLOAD=ON
ninja
ctest -V
```

**Test System**
 - BB5

**Use certain branches for the SimulationStack CI**

CI_BRANCHES:NEURON_BRANCH=master,

### CODE REVIEW

File: coreneuron/utils/randoms/nrnran123.cu [hunk 1]
```diff
@@ -76,23 +76,33 @@ using random123_allocator = coreneuron::unified_allocator<coreneuron::nrnran123_
...
 // __managed__ here, but unfortunately that does not work on machines that do
 // not have a GPU.
-#ifdef __CUDACC__
+#if defined(__CUDACC__) && !defined(CORENEURON_PREFER_OPENMP_OFFLOAD)
 CORENRN_DEVICE philox4x32_key_t* g_k_dev;
 #endif
 
 OMP_Mutex g_instance_count_mutex;
+
+nrn_pragma_omp(declare target)
```
Comment (olupton) [inline coreneuron/utils/randoms/nrnran123.cu:98]: I don't think we need `g_instance_count` on the GPU?

File: coreneuron/utils/randoms/nrnran123.cu [hunk 2]
```diff
@@ -152,7 +152,7 @@ nrn_pragma_omp(declare target)
 /** @brief Provide a helper function in global namespace that is declared target for OpenMP
  * offloading to function correctly with NVHPC
  */
-philox4x32_ctr_t philox4x32_helper(coreneuron::nrnran123_State* s) {
+CORENRN_HOST_DEVICE philox4x32_ctr_t philox4x32_helper(coreneuron::nrnran123_State* s) {
```
Comment (ohm314) [inline coreneuron/utils/randoms/nrnran123.cu:155]: ah riiight, i missed this!

File: coreneuron/utils/randoms/nrnran123.cu [hunk 3]
```diff
@@ -95,9 +95,7 @@ CORENRN_DEVICE philox4x32_key_t* g_k_dev;
 
 OMP_Mutex g_instance_count_mutex;
 
-nrn_pragma_omp(declare target)
```
Comment (ohm314) [inline coreneuron/utils/randoms/nrnran123.cu]: makes sense - I think that was left from an earlier attempt where the functions using this variable were also annoted.

File: coreneuron/CMakeLists.txt [hunk 4]
```diff
@@ -126,11 +126,16 @@ if(CORENRN_ENABLE_GPU)
   # (https://cmake.org/cmake/help/latest/policy/CMP0119.html), so using a single .cu file and
   # setting LANGUAGE=CXX in non-GPU builds does not work.
   list(APPEND CORENEURON_CODE_FILES ${CORENEURON_CUDA_FILES})
-  if(CORENRN_ENABLE_OPENMP_OFFLOAD)
+  if(CORENRN_ACCELERATOR_OFFLOAD STREQUAL "OpenMP")
```
Comment (ohm314) [inline coreneuron/CMakeLists.txt:129]: I was considering to implement it like this but thought that above option would be equivalent. Thanks for noticing this didn't actually produce the correct code.

**Comment by ohm314** [comment]
image

This is the channel benchmark with `--tstop=10`, on top is this PR, below the `master` branch. Profiler was run like this:

```
nsys profile --env-var NSYS_NVTX_PROFILER_REGISTER_ONLY=0,CALI_CONFIG=nvtx,OMP_NUM_THREADS=1 --trace=cuda,openacc,openmp -o corenrn_nsys_nvprof_baseline --capture-range=nvtx --nvtx-capture=main  x86_64/special-core --datpath=/nersc-gpu-hackathon-dec-2021/channel-benchmark-all-440-cells-2-ranks --mpi --gpu --cell-permute=2 --tstop=10
```

The two runs seem to perform about the same and the profile shows nothing unexpected for the new code.

**Review by olupton — APPROVED** [review]

**Pull request merged**